### PR TITLE
🥇 Refactor Collections and Observables

### PR DESCRIPTION
**Pull request by RomaricMourgues**

### CODE REVIEW

File: twake/frontend/src/app/services/Collections2/Resource.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    this._data = { id: this.genId(), ...data };
+  }
+
+  private genId() {
+    function s4() {
+      return Math.floor((1 + Math.random()) * 0x10000)
+        .toString(16)
+        .substring(1);
+    }
+    return 'tmp:' + s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
```
Comment (chamerling) [inline twake/frontend/src/app/services/Collections2/Resource.ts:20]: Can't we use an uuid lib for this?

File: twake/frontend/src/app/services/Collections2/__tests__/CollectionStorage.test.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
+import CollectionStorage from '../CollectionStorage';
+
+test('Mongo collection upsert and remove', async () => {
```
Comment (chamerling) [inline twake/frontend/src/app/services/Collections2/__tests__/CollectionStorage.test.ts:3]: You must split your tests

**Review by chamerling — COMMENTED** [review]
1. I think we can really simplify the code by using async/await or by cleaning the way Promise are called
2. I really think that the Collection lib must be an independent library, so that I can use it in any project

File: twake/frontend/src/app/services/Collections2/CollectionStorage.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  static upsert(path: string, item: any): Promise<any> {
+    return new Promise((resolve, reject) => {
+      if (!item.id) {
+        reject('Every resources must contain an id');
+        return;
+      }
+      CollectionStorage.addCollection(path);
+      CollectionStorage.find(path, { id: item.id })
+        .then(mongoItems => {
+          if (mongoItems.length === 1) {
```
[the remark this answers is not in the document]
Comment (RomaricMourgues) [reply twake/frontend/src/app/services/Collections2/CollectionStorage.ts:28]: Where did you find this ? In my docs mongoItems always contain a list (empty or with items but a list)

[the remark this answers is not in the document]
Comment (chamerling) [reply twake/frontend/src/app/services/Collections2/CollectionStorage.ts:28]: Does this always return an array?

```js
CollectionStorage.getMongoDb()
        .collections[path].find(filters, options)
        .fetch(resolve, reject);
```

?

[the remark this answers is not in the document]
Comment (RomaricMourgues) [reply twake/frontend/src/app/services/Collections2/CollectionStorage.ts:42]: Yes, fetch definition is 
```
(method) fetch(success: (items: any[]) => void, error: (err: any) => void): void
```
So array or error (no error if empty list)

[the remark this answers is not in the document]
Comment (RomaricMourgues) [reply twake/frontend/src/app/services/Collections2/CollectionStorage.ts:42]: Currently tested in CollectionsStorage.test.ts and it seems to works like this (empty lists and not null result)

**Pull request merged**